### ISSUE TITLE
flow_from_directory doesn't create enough batches

### ISSUE THREAD

**Issue by LucasColas** [tensorflow/tensorflow#47776]
Hello I have an issue with flow_from_directory to yield enough data.

**System information**

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- TensorFlow version (use command below): TF 2.5
- Python version: 3.7

- CUDA/cuDNN version: CUDA 11.2
- GPU model and memory: GTX 1660 TI

**Describe the current behavior**
I wanted to yield data with flow_from_directory method. But it didn't work well.

**Describe the expected behavior**
I expected to have enough batches for my training.

**Standalone code to reproduce the issue**
` from tensorflow.keras import layers
from tensorflow.keras import models
from tensorflow.keras import optimizers
from tensorflow.keras.preprocessing.image import ImageDataGenerator

import matplotlib.pyplot as plt

import os

new_DS = r'C:\Users\lucas\Projets Code\ConvNet Dogs And Cats\cats_and_dogs_small'
train_folder = os.path.join(new_DS, 'train')
test_folder = os.path.join(new_DS, 'test')

train_datagen = ImageDataGenerator(rescale=1./255, rotation_range=40,width_shift_range=0.2,height_shift_range=0.2, shear_range=0.2, zoom_range=0.2, horizontal_flip=True)
validation_folder = os.path.join(new_DS, 'validation')
test_datagen = ImageDataGenerator(rescale=1./255)

train_generator = train_datagen.flow_from_directory(train_folder,target_size=(150,150),batch_size=32, class_mode='binary')

validation_generator = test_datagen.flow_from_directory(validation_folder, target_size=(150,150),batch_size=32, class_mode='binary')

model = models.Sequential()
model.add(layers.Conv2D(32, (3,3), activation='relu',input_shape=(150,150,3)))
model.add(layers.MaxPooling2D((2,2)))
model.add(layers.Conv2D(64, (3,3), activation='relu'))
model.add(layers.MaxPooling2D((2,2)))
model.add(layers.Conv2D(128, (3,3), activation='relu'))
model.add(layers.MaxPooling2D((2,2)))
model.add(layers.Conv2D(128, (3,3), activation='relu'))
model.add(layers.MaxPooling2D((2,2)))
model.add(layers.Flatten())
model.add(layers.Dropout(0.5))
model.add(layers.Dense(512, activation='relu')) #provides learning features from all the combinations
                                                #of the features of the previous layer
model.add(layers.Dense(1,activation='sigmoid'))

model.compile(loss='binary_crossentropy', optimizer=optimizers.RMSprop(lr=1e-4), metrics=['acc'])

history = model.fit(train_generator, steps_per_epoch=100, epochs=100, validation_data=validation_generator,validation_steps=50)
model.save('cats_and_dogs_2.h5')
acc = history.history['acc']
val_acc = history.history['val_acc']
loss = history.history['loss']
val_loss = history.history['val_loss']

epochs = range(1,len(acc)+1)

plt.plot(epochs,acc,'bo', label='Training acc')
plt.plot(epochs, val_acc,'b', label='Validation acc')
plt.title('Training and validation accuracy')
plt.legend()

plt.figure()

plt.plot(epochs, loss, 'bo', label='Training loss')
plt.plot(epochs, val_loss, 'b', label='Validation loss')
plt.title('Training and validation loss')
plt.legend()

plt.show()
 `

**Other info / logs** Include any logs or source code that would be helpful to
Here is what I got in my terminal : 

> To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-03-13 06:51:31.781669: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1304] Device interconnect StreamExecutor with strength 1 edge matrix:
2021-03-13 06:51:31.792390: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1310]
2021-03-13 06:51:32.680059: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:176] None of the MLIR Optimization Passes are enabled (registered 2)
Epoch 1/100
 63/100 [=================>............] - ETA: 22s - loss: 0.7003 - acc: 0.4875WARNING:tensorflow:Your input ran out of data; interrupting training. Make sure that your dataset or generator can generate at least `steps_per_epoch * epochs` batches (in this case, 10000 batches). You may need to use the repeat() function when building your dataset.
WARNING:tensorflow:Your input ran out of data; interrupting training. Make sure that your dataset or generator can generate at least `steps_per_epoch * epochs` batches (in this case, 50 batches). You may need to use the repeat() function when building your dataset.
100/100 [==============================] - 44s 418ms/step - loss: 0.6964 - acc: 0.4993 - val_loss: 0.7465 - val_acc: 0.5000

**Comment by AdityaKane2001** [comment]
@LucasColas 
Note that `steps_per_epoch` is the number of batches to consume in one epoch. You can omit it to enable the model to go through the whole dataset once every epoch. If `steps_per_epoch` is mentioned, make sure that you have atleast `batch_size*steps_per_epoch*epochs` number of individual samples in your dataset.

Hope this helps.

**Comment by LucasColas** [comment]
I use `flow_from_directory` so it generates batches of augmented data...

**Comment by amahendrakar** [comment]
@LucasColas,
In order to expedite the trouble-shooting process, could you please provide the dataset you are using and all the files required to run the code. Thanks!

**Comment by LucasColas** [comment]
How can I send the dataset ? It's a very large folder

**Comment by LucasColas** [comment]
> 
> 
> @LucasColas
> Note that `steps_per_epoch` is the number of batches to consume in one epoch. You can omit it to enable the model to go through the whole dataset once every epoch. If `steps_per_epoch` is mentioned, make sure that you have atleast `batch_size*steps_per_epoch*epochs` number of individual samples in your dataset.
> 
> Hope this helps.

But `flow_from_directory` yields data indefinitely, right ?

**Comment by AdityaKane2001** [comment]
@LucasColas 
Yes, it is infinite, but its termination is handled by `model.fit()` and `steps_per_epoch` and `epochs` parameters. It throws an error when the data is iterated once. For more flexibility, I recommend using `tf.data.Dataset`.

**Comment by LucasColas** [comment]
It's not really useful, is it ?

**Comment by amahendrakar** [comment]
> How can I send the dataset ? It's a very large folder

@LucasColas,
If it's a publicly available dataset you can the link with us or you can also upload the dataset to Google Drive and share the folder with us. Thanks!
